### ISSUE TITLE
Version History for Devices

### ISSUE THREAD

**Issue by joepavitt** [FlowFuse/flowfuse#4873]
### Description

Follow on item from: https://github.com/FlowFuse/flowfuse/pull/4509

Create the Version History view for Devices

```[tasklist]
- [ ] API
- [ ] UI
```

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

**Comment by knolleary** [comment]
I think we move foward with option 1 to get the basic api and UI in place. We then formulate a plan to address 2/3 as a follow on (rather than try to solve them before making progress on 1).

I don't have an immediate suggestion on solving 2/3 - will need to digest and think about it some more.

**Comment by joepavitt** [comment]
The events we need to understand are always going to be `flows.set`, but it's important to communicate to the user _why_ this happened. We have a few ways by which flows can be deployed to a Remote Instance:

- Instance is in Dev Mode, and the user has explicitly clicked "Deploy" inside the editor
- DevOps Pipeline: Pushed from one stage to another
- Instance Target Snapshot: Remote Instance is bound to a Hosted Instance, and the target snapshot has been set and rolled out to the Remote Instance.
- Snapshot Rollback: user has chosen an old snapshot to re-deploy to the Instance

If we have an event for each of these four, then we are golden.

You can read the events that were used for Hosted Instances here: https://github.com/FlowFuse/flowfuse/issues/4424#issuecomment-2345743778

**Comment by joepavitt** [comment]
The version history view should only be showing the following events:

- Manual deploy - `flows.set`
- Pipeline - `application.pipeline.stage.deployed` - this currently triggers at application level, and not necessarily reliant, needs further investigation
- Instance Target Snapshot - ``
- Snapshot Rollback: 
- Snapshot Created: 

Please can you identify the gaps here and detail explicitly which audit log events correlate to these things happening on the Device.

**Comment by cstns** [comment]
> The version history view should only be showing the following events:
> 
>     * Manual deploy - `flows.set`
> 
>     * Pipeline - `application.pipeline.stage.deployed` - this currently triggers at application level, and not necessarily reliant, needs further investigation
> 
>     * Instance Target Snapshot - ``
> 
>     * Snapshot Rollback: `application.device.snapshot.device-target-set`
> 
>     * Snapshot Created: ``
> 
> 
> Please can you identify the gaps here and detail explicitly which audit log events correlate to these things happening on the Device.

Current state of affairs:
- **manual deploy:**
	- event: `flows.set`
	- reliable, direct correlation with the device

- **deploying through a pipeline:**
	- events: `application.device.snapshot.device-target-set` or `team.device.updated`
	- the events are bound to the application and team respectively

- **deploying through an instance**
	- event: `project.snapshot.device-target-set`
	- the event is bound to the instance

- **snapshot rollback:**
	- event: `team.device.updated` or `application.device.snapshot.device-target-set`
	- can only be triggered when the device is assigned to an application	
	- the events are bound to the application and team respectively

- **spanshot created:**
	- event: `application.device.snapshot.created`
	- can only be triggered when the device is assigned to an application
	- the event is bound to the application


Other than the manual deploy event we don't have anything registered on the device itself. For an mvp we could at least insert a device audit log event starting from this moment onward when these project/team/applications events are triggered so users can at least actively keep track of their device flow changes.

Backtracking these would be worth it on the long run but it would need careful implementation.

**Comment by joepavitt** [comment]
> We could at least insert a device audit log event starting from this moment onward when these project/team/applications events are triggered so users can at least actively keep track of their device flow changes.

Agreed, but separate PR please as a follow up as to not block this.

**Comment by joepavitt** [comment]
Closing this out as the primary task has been completed. Follow up item with #5111 has been created for next Milestone
